segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,263 @@
+     */
+    public GarbageCollector(ChunkedSegmentStorage chunkedSegmentStorage, ChunkedSegmentStorageConfig config) {
+        super("io.pravega.segmentstore.storage.chunklayer.GarbageCollector", (ScheduledExecutorService) chunkedSegmentStorage.getExecutor());
+        this.chunkedSegmentStorage = Preconditions.checkNotNull(chunkedSegmentStorage, "chunkedSegmentStorage");
+        this.config = Preconditions.checkNotNull(config, "config");
+    }
+
+    /**
+     * Initializes this instance.
+     */
+    public void initialize() {
+        Services.startAsync(this, this.executor);
+    }
+
+    /**
+     * Gets a value indicating how much to wait for the service to shut down, before failing it.
+     *
+     * @return The Duration.
+     */
+    @Override
+    protected Duration getShutdownTimeout() {
+        return null;
+    }
+
+    /**
+     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     *
+     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
+     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     */
+    @Override
+    protected CompletableFuture<Void> doRun() {
+        loopFuture = Futures.loop(
+                this::canRun,
+                () -> CompletableFuture.completedFuture(null)
+                        .thenComposeAsync( v -> deleteGarbage(true, config.getGarbageCollectionConcurrency()), executor)
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{} Error during run.", chunkedSegmentStorage.getLogPrefix(), ex);
+                            }
+                            return null;
+                        }, executor),
+                executor);
+        return loopFuture;
+    }
+
+    private boolean canRun() {
+        return isRunning() && getStopException() == null && !closed.get();
+    }
+
+    /**
+     * Sets whether background cleanup is suspended or not.
+     *
+     * @param value Boolean indicating whether to suspend background processing or not.
+     */
+    void setSuspended(boolean value) {
+        suspended.set(value);
+    }
+
+    /**
+     * collect the garbage chunks.
+     *
+     * @param chunksToDelete List of chunks to delete.
+     */
+    void addToGarbage(Collection<String> chunksToDelete) {
+        val currentTime = System.currentTimeMillis();
+
+        if (queueSize.get() < config.getGarbageCollectionMaxQueueSize()) {
+            chunksToDelete.forEach(chunkToDelete -> garbageChunks.add(new GarbageChunkInfo(chunkToDelete, currentTime)));
+            queueSize.incrementAndGet();
+        } else {
+            for (val chunkToDelete : chunksToDelete) {
+                log.warn("{} deleteGarbage - Queue full. Could not delete garbage. chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
```

Comment:
This is dangerous and can flood all the Segment Store logs, leading to lose logs due to their sheer size and/or making debug much harder. Please, add a single log line with the length of `chunksToDelete` if you want.

Reply:
ok . made it a debug.